File: code/go/ch11/src/luago/state/closure.go
package state

import . "luago/api"
import "luago/binchunk"

type closure struct {
	proto  *binchunk.Prototype // lua closure
	goFunc GoFunction          // go closure
	upvals []*luaValue
}

func newLuaClosure(proto *binchunk.Prototype) *closure {
	upvals := make([]*luaValue, len(proto.Upvalues))
	return &closure{
		proto:  proto,
		upvals: upvals,
	}
}

func newGoClosure(f GoFunction, nUpvals int) *closure {
	upvals := make([]*luaValue, nUpvals)
	return &closure{
		goFunc: f,
		upvals: upvals,
	}
}

File: code/java/ch09/src/main/java/com/github/zxh0/luago/state/Closure.java
package com.github.zxh0.luago.state;

import com.github.zxh0.luago.binchunk.Prototype;
import lombok.Getter;

@Getter
class Closure {

    final Prototype proto;

    Closure(Prototype proto) {
        this.proto = proto;
    }

}
